## Project UNDP ACCELERATOR LABS

In [42]:
# Importing Packages
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from googletrans import Translator
import translators as ts

%matplotlib inline

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)


In [305]:
from googletrans import Translator

In [3]:
# Loading data
data = pd.read_excel("Viz4SocialGood_submissionV1.xlsx", sheet_name = "Energy_sol_data")

In [4]:
energy_data = data.copy()

In [5]:
energy_data.head()

contribution_date  \
0  Sun May 30 2021 13:48:32 GMT+0000 (Coordinated...   
1  Mon Jun 28 2021 13:56:30 GMT+0000 (Coordinated...   
2  Mon Jun 28 2021 13:32:36 GMT+0000 (Coordinated...   
3  Thu Nov 12 2020 06:03:43 GMT+0000 (Coordinated...   
4  Wed Jul 21 2021 10:12:39 GMT+0000 (Coordinated...   

                   new_date    id          Energy source  \
0  Sun May 30 2021 13:48:32  3100                Thermal   
1  Mon Jun 28 2021 13:56:30  3177                   Wind   
2  Mon Jun 28 2021 13:32:36  3176                  Hydro   
3  Thu Nov 12 2020 06:03:43  2261      Renewable general   
4  Wed Jul 21 2021 10:12:39  3236  Household application   

  Clean cooking application (yes)                                  title  \
0                             NaN                               Eco Char   
1                             NaN                   Wind power generator   
2                             NaN                       Hybrid Generator   
3                             NaN  Electric Power Microgeneration System   
4                               x                           Solar Cooker   

           Mapper Contributor anonymized  \
0  AccLab Algeria          Contributor 1   
1   AccLab Angola          Contributor 2   
2   AccLab Angola          Contributor 3   
3   AccLab Angola          Contributor 4   
4   AccLab Angola          Contributor 5   

  What is the purpose of the solution? (brief problem & solution description)  \
0  We upcycle wastes such as food organic wastes ...                            
1  The system to obtain electrical energy through...                            
2  The production of fuel from water is done thro...                            
3  The electric power microgeneration system is d...                            
4   Solar Cooker\n\n\n\t\t\t\n\n\n\t\t\t\n\n\n\t\t\t                            

  Please insert a link to the solution  \
0                                  NaN   
1                                  NaN   
2                                  NaN   
3                                  NaN   
4                                  NaN   

  This solution is Do it Yourself / open source  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   

  This solution is protected by Intellectual Property  \
0                                                NaN    
1                                                NaN    
2                                                NaN    
3                                                NaN    
4                                                NaN    

  The solution holder is able to train others (including end-users) in using or replicating the solution  \
0                                                NaN                                                       
1                                                NaN                                                       
2                                                NaN                                                       
3                                                NaN                                                       
4                                                NaN                                                       

  What is the unit cost of this Solution along with any additional cost for maintenance and training?  \
0                                                NaN                                                    
1                                                NaN                                                    
2                                                NaN                                                    
3                                                NaN                                                    
4                                                Na

In [6]:
energy_data['new_date'] = pd.to_datetime(energy_data['new_date'])
energy_data.sample(5)

contribution_date            new_date  \
260  Tue Oct 06 2020 16:26:27 GMT+0000 (Coordinated... 2020-10-06 16:26:27   
336  Thu Jun 09 2022 08:45:18 GMT+0000 (Coordinated... 2022-06-09 08:45:18   
321  Tue May 31 2022 11:23:03 GMT+0000 (Coordinated... 2022-05-31 11:23:03   
291  Sat Aug 29 2020 13:51:00 GMT+0000 (Coordinated... 2020-08-29 13:51:00   
358  Thu Jun 09 2022 11:02:12 GMT+0000 (Coordinated... 2022-06-09 11:02:12   

       id          Energy source Clean cooking application (yes)  \
260  1674      Renewable general                             NaN   
336  5061  Household application                               x   
321  4948  Household application                               x   
291   109                Thermal                             NaN   
358  5094  Household application                               x   

                                                 title  \
260                   Design Cube Eco-Friendly Housing   
336                                       Rocket Stove   
321                           Salan Mini Biomass Stove   
291  United Innovations Development Center- Making ...   
358                                Modified Wood Stove   

                        Mapper Contributor anonymized  \
260        AccLab Sierra Leone        Contributor 261   
336  Honey Bee Network (India)        Contributor 337   
321  Honey Bee Network (India)        Contributor 322   
291              AccLab Uganda        Contributor 292   
358  Honey Bee Network (India)        Contributor 359   

    What is the purpose of the solution? (brief problem & solution description)  \
260  Design Cube are into transforming of shipping ...                            
336  The Innovator has developed Rocket Stove a mod...                            
321  The Innovator faced problems with the traditio...                            
291  Conversion of Biomass Waste into clean energy ...                            
358  The Innovator noticed the workers taking long ...                            

    Please insert a link to the solution  \
260                                  NaN   
336                                  NaN   
321                                  NaN   
291                                  NaN   
358                                  NaN   

    This solution is Do it Yourself / open source  \
260                                           NaN   
336                                           NaN   
321                                           NaN   
291                                           NaN   
358                                           NaN   

    This solution is protected by Intellectual Property  \
260                                                NaN    
336                                                NaN    
321                                                NaN    
291                                                NaN    
358                                                NaN    

    The solution holder is able to train others (including end-users) in using or replicating the solution  \
260                                                NaN                                                       
336                                                NaN                                                       
321                                                NaN                                                       
291                                                NaN                                                       
358                                                NaN                                                       

    What is the unit cost of this Solution along with any additional cost for maintenance and training?  \
260                                                NaN                                                    
336                                                NaN                                                    
321                                            

In [7]:
energy_data.columns

Index(['contribution_date', 'new_date', 'id', 'Energy source',
       'Clean cooking application (yes)', 'title', 'Mapper',
       'Contributor anonymized',
       'What is the purpose of the solution? (brief problem & solution description)',
       'Please insert a link to the solution',
       'This solution is Do it Yourself / open source',
       'This solution is protected by Intellectual Property',
       'The solution holder is able to train others (including end-users) in using or replicating the solution',
       'What is the unit cost of this Solution along with any additional cost for maintenance and training?',
       'This solution is a Prototype', 'This solution is a Product',
       'If this solution is a product, is it available in market?',
       'If this solution is a product, does advance order has to be given?',
       'What is the Technological Readiness Level (TRL) of this solution?',
       'How much has this solution already been diffused? Is there potential fe

In [8]:
energy_data.columns
energy_data.drop('contribution_date', inplace=True, axis=1)

In [9]:
# Here, column names have been renamed to shorter but descriptive forms.
# To save time, the column names were changed to title case so we don't have to rename them in our charts.

energy_data.rename(columns = {'new_date' : 'Date', 
        'id': 'ID',
        'Energy source': 'Energy Source',
       'Clean cooking application (yes)' : 'Clean Cooking',
       'title': 'Project Title',
       'Contributor anonymized' : 'Contributor',
       'What is the purpose of the solution? (brief problem & solution description)' : 'Purpose of Solution',
       'Please insert a link to the solution' : 'Link',
       'This solution is Do it Yourself / open source' : 'Open Source',
       'This solution is protected by Intellectual Property' : 'IP Protected',
       'The solution holder is able to train others (including end-users) in using or replicating the solution' : 'Can Train Others',
       'What is the unit cost of this Solution along with any additional cost for maintenance and training?' : 'Unit Cost in USD',
       'This solution is a Prototype' : 'Prototype', 
        'This solution is a Product' : 'Product',
       'If this solution is a product, is it available in market?' : 'Market Ready',
       'If this solution is a product, does advance order has to be given?' : 'Requires Advanced Order',
       'What is the Technological Readiness Level (TRL) of this solution?' : 'Tech Readiness Level',
       'How much has this solution already been diffused? Is there potential feedback from end-users available?' : 'User Feedback',
       'Please upload a link of end-user feedback' : 'Feedback Link',
       'Are there any efficiency benchmarks for this solution (eg. how much energy does it save; how much cheaper does it produce energy than current market rates/ current household expenditure / cost per kW h)?' : 'Efficiency Benchmark',
       'Are there any other potential bottlenecks affecting cross-border or in country diffusion of this solution?': 'Potential Bottlenecks',
       'What Sustainable Development Goal is this Solution addressing? Tag 1' : 'SDG Solution1',
       'What Sustainable Development Goal is this Solution addressing? Tag 2' : 'SDG Solution2',
       'What Sustainable Development Goal is this Solution addressing? Tag 3' : 'SDG Solution3',
       'What Sustainable Development Goal is this Solution addressing? Tag 4' : 'SDG Solution4',
       'What Sustainable Development Goal is this Solution addressing? Tag 5' : 'SDG Solution5',
       'What thematic tags apply to this solution? Tag 1' : 'Thematic Tag1',
       'What thematic tags apply to this solution? Tag 2' : 'Thematic Tag2',
       'What thematic tags apply to this solution? Tag 3' : 'Thematic Tag3',
       'What thematic tags apply to this solution? Tag 4' : 'Thematic Tag4',
       'What thematic tags apply to this solution? Tag 5' : 'Thematic Tag5', 
        'Latitude ' : 'Lat',
       'Longitude ' : 'Long'}, inplace = True)

In [10]:
energy_data.head()

Date    ID          Energy Source Clean Cooking  \
0 2021-05-30 13:48:32  3100                Thermal           NaN   
1 2021-06-28 13:56:30  3177                   Wind           NaN   
2 2021-06-28 13:32:36  3176                  Hydro           NaN   
3 2020-11-12 06:03:43  2261      Renewable general           NaN   
4 2021-07-21 10:12:39  3236  Household application             x   

                           Project Title          Mapper    Contributor  \
0                               Eco Char  AccLab Algeria  Contributor 1   
1                   Wind power generator   AccLab Angola  Contributor 2   
2                       Hybrid Generator   AccLab Angola  Contributor 3   
3  Electric Power Microgeneration System   AccLab Angola  Contributor 4   
4                           Solar Cooker   AccLab Angola  Contributor 5   

                                 Purpose of Solution Link Open Source  \
0  We upcycle wastes such as food organic wastes ...  NaN         NaN   
1  The system to obtain electrical energy through...  NaN         NaN   
2  The production of fuel from water is done thro...  NaN         NaN   
3  The electric power microgeneration system is d...  NaN         NaN   
4   Solar Cooker\n\n\n\t\t\t\n\n\n\t\t\t\n\n\n\t\t\t  NaN         NaN   

  IP Protected Can Train Others Unit Cost in USD Prototype Product  \
0          NaN              NaN              NaN       NaN     NaN   
1          NaN              NaN              NaN       NaN     NaN   
2          NaN              NaN              NaN       NaN     NaN   
3          NaN              NaN              NaN       NaN     NaN   
4          NaN              NaN              NaN       NaN     NaN   

  Market Ready Requires Advanced Order Tech Readiness Level User Feedback  \
0          NaN                     NaN                  NaN           NaN   
1          NaN                     NaN                  NaN           NaN   
2          NaN                     NaN                  NaN           NaN   
3          NaN                     NaN                  NaN           NaN   
4          NaN                     NaN                  NaN           NaN   

  Feedback Link Efficiency Benchmark Potential Bottlenecks  SDG Solution1  \
0           NaN                  NaN                   NaN            1.0   
1           NaN                  NaN                   NaN            1.0   
2           NaN                  NaN                   NaN           10.0   
3           NaN                  NaN                   NaN            1.0   
4           NaN                  NaN                   NaN            7.0   

   SDG Solution2  SDG Solution3  SDG Solution4  SDG Solution5  \
0            6.0           13.0            NaN            NaN   
1            7.0           13.0            NaN            NaN   
2           12.0           13.0            NaN            NaN   
3            7.0            9.0           10.0           11.0   
4           12.0           13.0            NaN            NaN   

           Thematic Tag1          Thematic Tag2          Thematic Tag3  \
0  food waste management                  youth                 coffee   
1     alternative energy           clean energy         climate change   
2                    NaN     alternative energy            electricity   
3           clean energy  community empowerment  rural electrification   
4             innovation     polution reduction           solar energy   

           Thematic Tag4 Thematic Tag5         Lat        Long        image_1  \
0                 olives           NaN  36.7607349   3.0261856            NaN   
1  community empowerment           NaN  -12.833226  13.6010742   file: img-10   
2     polution reduction           NaN  -8.5158356  16.0817871  file: img-121   
3         social justice       welfare  -12.776115   15.687688  file: img-230   
4                    NaN           NaN   -12.86536  13.8098145  file: img-305   

  image_2 image_3 image_4 image_5 image_6 image_

In [11]:
energy_data['Unit Cost in USD'].value_counts()

Undefined                                                                                                                                                                                                                                                                                         2
Unsure.                                                                                                                                                                                                                                                                                           2
20 USD                                                                                                                                                                                                                                                                                            2
1200 BDT per unit                                                                                                           

In [12]:
# Load statistics of data
print(energy_data.shape)
energy_data.info()

(359, 43)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359 entries, 0 to 358
Data columns (total 43 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Date                     357 non-null    datetime64[ns]
 1   ID                       359 non-null    int64         
 2   Energy Source            355 non-null    object        
 3   Clean Cooking            82 non-null     object        
 4   Project Title            359 non-null    object        
 5   Mapper                   359 non-null    object        
 6   Contributor              359 non-null    object        
 7   Purpose of Solution      347 non-null    object        
 8   Link                     126 non-null    object        
 9   Open Source              100 non-null    object        
 10  IP Protected             41 non-null     object        
 11  Can Train Others         127 non-null    object        
 12  Unit Cost in USD         7

### About the data
The data consists of 359 entries for 44 categories.

### Data quality issues
- There are lots of Null values with few incorrect data types
- Invalid data in the lat and long columns [xxxx, xxxxx]
- Non-descriptive values 'x' in columns: ['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others', 'Prototype', 'Product', 'Market Ready', 'Requires Advanced Order']
- Columns are in wrong datatypes [ID, ]
- The coulmn 'Unit Cost in USD' consists of a lot of currencies, numbers and texts.
- Some textx are not in English

### Tidiness Issues
- Some columns are opposite of the other, thus should form one column. Affected columns include 
[Prototype, Product	], [Solution1	SDG Solution2	SDG Solution3	SDG Solution4	SDG Solution5], [Thematic Tag1	Thematic Tag2	Thematic Tag3	Thematic Tag4	Thematic Tag5]

In [13]:
energy_data.sample(3)

Date    ID          Energy Source Clean Cooking  \
356 2022-06-09 10:52:58  5091  Household application             x   
43  2022-05-17 09:55:53  4777                  Solar           NaN   
293 2020-07-24 10:18:16    43  Household application             x   

                                         Project Title  \
356                             Modified Cooking Stove   
43        Solar powered water filtration in Bangladesh   
293  Fireless Cookers for energy conservation in Ug...   

                        Mapper      Contributor  \
356  Honey Bee Network (India)  Contributor 357   
43           AccLab Bangladesh   Contributor 44   
293              AccLab Uganda  Contributor 294   

                                   Purpose of Solution Link Open Source  \
356  The Innovatorhas developed a modified cooking ...  NaN           x   
43   In Bangladesh, access to water is a severe cha...  NaN           x   
293  A fireless cooker is an insulated basket which...  NaN         NaN   

    IP Protected Can Train Others Unit Cost in USD Prototype Product  \
356          NaN              NaN              NaN         x     NaN   
43           NaN              NaN              NaN       NaN     NaN   
293          NaN              NaN              NaN       NaN     NaN   

    Market Ready Requires Advanced Order Tech Readiness Level User Feedback  \
356          NaN                     NaN                  NaN           NaN   
43           NaN                     NaN                Ready           NaN   
293          NaN                     NaN                  NaN           NaN   

    Feedback Link Efficiency Benchmark Potential Bottlenecks  SDG Solution1  \
356           NaN                  NaN                   NaN            7.0   
43            NaN                  NaN                   NaN            NaN   
293           NaN                  NaN                   NaN            2.0   

     SDG Solution2  SDG Solution3  SDG Solution4  SDG Solution5  \
356            NaN            NaN            NaN            NaN   
43             NaN            NaN            NaN            NaN   
293            7.0           13.0            NaN            NaN   

      Thematic Tag1  Thematic Tag2         Thematic Tag3 Thematic Tag4  \
356   clean cooking        cooking  environment friendly         stove   
43              NaN            NaN                   NaN           NaN   
293  climate change  deforestation                energy           NaN   

    Thematic Tag5         Lat        Long        image_1       image_2  \
356           NaN   24.818124  85.5185562            NaN           NaN   
43            NaN  22.9319753  91.4967701  file: img-537           NaN   
293           NaN  0.34158624  32.6248741   file: img-38  file: img-39   

    image_3 image_4 image_5 image_6 image_7 image_8 image_9  
356     NaN     NaN     NaN     NaN     NaN     NaN     NaN  
43      NaN     NaN     NaN     NaN     NaN     NaN     NaN  
293     NaN     NaN     NaN     NaN     NaN     NaN     NaN

In [14]:
col_index = pd.DataFrame(np.where(energy_data.eq('x'))[1], columns=['col_index'])

In [15]:
col_index.groupby('col_index')['col_index'].count()

col_index
3      82
9     100
10     41
11    127
13     69
14    124
15     64
16     61
Name: col_index, dtype: int64

In [16]:
energy_data.columns[[3,9,10,11,13,14,15,16]]

Index(['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
       'Prototype', 'Product', 'Market Ready', 'Requires Advanced Order'],
      dtype='object')

In [17]:
energy_data['Clean Cooking'].unique()

array([nan, 'x'], dtype=object)

In [18]:
# EDITED
# Replaced x with other values (instead of 1) in below columns instead of 0. This will maake it more interpretable by others

# Columns that will be combined subsequently
energy_data['Open Source'] = energy_data['Open Source'].replace('x', 'Open Source')
energy_data['IP Protected'] = energy_data['IP Protected'].replace('x', 'IP Protected')
energy_data['Prototype'] = energy_data['Prototype'].replace('x', 'Prototype')
energy_data['Product'] = energy_data['Product'].replace('x', 'Product')

In [19]:
# Other columns that do not need to be combined
# For these ones, we shall replace x with True and NaN with False
energy_data[['Clean Cooking', 'Can Train Others', 
             'Market Ready', 'Requires Advanced Order']] = energy_data[['Clean Cooking', 'Can Train Others', 
                                                                        'Market Ready', 'Requires Advanced Order']].replace('x', True)
energy_data[['Clean Cooking', 'Can Train Others', 
             'Market Ready', 'Requires Advanced Order']] = energy_data[['Clean Cooking', 'Can Train Others', 
             'Market Ready', 'Requires Advanced Order']].fillna(False)

In [20]:
# energy_data[['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
#        'Prototype', 'Product', 'Market Ready',
#        'Requires Advanced Order']] = energy_data[['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
#        'Prototype', 'Product', 'Market Ready',
#        'Requires Advanced Order']].replace('x', True)

In [21]:
energy_data[['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
       'Prototype', 'Product', 'Market Ready', 'Requires Advanced Order']].sample(10)

Clean Cooking  Open Source  IP Protected  Can Train Others  Prototype  \
251          False  Open Source           NaN              True        NaN   
157          False          NaN           NaN             False        NaN   
189          False          NaN           NaN             False        NaN   
249          False          NaN           NaN              True        NaN   
266          False  Open Source           NaN              True        NaN   
263           True          NaN           NaN             False        NaN   
203          False          NaN           NaN             False        NaN   
125          False          NaN  IP Protected              True  Prototype   
273          False          NaN           NaN             False        NaN   
15            True  Open Source           NaN              True        NaN   

     Product  Market Ready  Requires Advanced Order  
251      NaN         False                    False  
157      NaN         False                    False  
189      NaN         False                    False  
249      NaN         False                    False  
266  Product         False                     True  
263      NaN         False                    False  
203      NaN         False                    False  
125      NaN         False                     True  
273      NaN         False                    False  
15   Product          True                    False

In [22]:
# No need replacing nan with false since we will eventually combine some of these columns into long forms
# energy_data[['clean cooking', 'opensource', 'ip_protection', 'replicated',
#        'prototype', 'product', 'market_availability',
#        'advanced_order_applicable']] = energy_data[['clean cooking', 'opensource', 'ip_protection', 'replicated',
#        'prototype', 'product', 'market_availability',
#        'advanced_order_applicable']].applymap(bool)

In [23]:
energy_data.sample()

Date   ID Energy Source  Clean Cooking  \
312 2020-09-18 12:06:38  933       Thermal          False   

                               Project Title           Mapper  \
312  Bio'plus briquettes made from bio-waste  AccLab Zimbabwe   

         Contributor                                Purpose of Solution  \
312  Contributor 313  These briquettes are made from agriculture was...   

                                 Link Open Source IP Protected  \
312  https://youtu.be/ESdN-HaW1JM\n\n         NaN          NaN   

     Can Train Others Unit Cost in USD Prototype Product  Market Ready  \
312             False              NaN       NaN     NaN         False   

     Requires Advanced Order Tech Readiness Level User Feedback Feedback Link  \
312                    False                  NaN           NaN           NaN   

    Efficiency Benchmark Potential Bottlenecks  SDG Solution1  SDG Solution2  \
312                  NaN                   NaN           12.0           13.0   

     SDG Solution3  SDG Solution4  SDG Solution5     Thematic Tag1  \
312            NaN            NaN            NaN  circular economy   

      Thematic Tag2                            Thematic Tag3 Thematic Tag4  \
312  climate change  community  involment, youth, employment        energy   

        Thematic Tag5         Lat        Long        image_1 image_2 image_3  \
312  waste management  -29.823593  27.2404083  file: img-192     NaN     NaN   

    image_4 image_5 image_6 image_7 image_8 image_9  
312     NaN     NaN     NaN     NaN     NaN     NaN

In [24]:
energy_data.head()

Date    ID          Energy Source  Clean Cooking  \
0 2021-05-30 13:48:32  3100                Thermal          False   
1 2021-06-28 13:56:30  3177                   Wind          False   
2 2021-06-28 13:32:36  3176                  Hydro          False   
3 2020-11-12 06:03:43  2261      Renewable general          False   
4 2021-07-21 10:12:39  3236  Household application           True   

                           Project Title          Mapper    Contributor  \
0                               Eco Char  AccLab Algeria  Contributor 1   
1                   Wind power generator   AccLab Angola  Contributor 2   
2                       Hybrid Generator   AccLab Angola  Contributor 3   
3  Electric Power Microgeneration System   AccLab Angola  Contributor 4   
4                           Solar Cooker   AccLab Angola  Contributor 5   

                                 Purpose of Solution Link Open Source  \
0  We upcycle wastes such as food organic wastes ...  NaN         NaN   
1  The system to obtain electrical energy through...  NaN         NaN   
2  The production of fuel from water is done thro...  NaN         NaN   
3  The electric power microgeneration system is d...  NaN         NaN   
4   Solar Cooker\n\n\n\t\t\t\n\n\n\t\t\t\n\n\n\t\t\t  NaN         NaN   

  IP Protected  Can Train Others Unit Cost in USD Prototype Product  \
0          NaN             False              NaN       NaN     NaN   
1          NaN             False              NaN       NaN     NaN   
2          NaN             False              NaN       NaN     NaN   
3          NaN             False              NaN       NaN     NaN   
4          NaN             False              NaN       NaN     NaN   

   Market Ready  Requires Advanced Order Tech Readiness Level User Feedback  \
0         False                    False                  NaN           NaN   
1         False                    False                  NaN           NaN   
2         False                    False                  NaN           NaN   
3         False                    False                  NaN           NaN   
4         False                    False                  NaN           NaN   

  Feedback Link Efficiency Benchmark Potential Bottlenecks  SDG Solution1  \
0           NaN                  NaN                   NaN            1.0   
1           NaN                  NaN                   NaN            1.0   
2           NaN                  NaN                   NaN           10.0   
3           NaN                  NaN                   NaN            1.0   
4           NaN                  NaN                   NaN            7.0   

   SDG Solution2  SDG Solution3  SDG Solution4  SDG Solution5  \
0            6.0           13.0            NaN            NaN   
1            7.0           13.0            NaN            NaN   
2           12.0           13.0            NaN            NaN   
3            7.0            9.0           10.0           11.0   
4           12.0           13.0            NaN            NaN   

           Thematic Tag1          Thematic Tag2          Thematic Tag3  \
0  food waste management                  youth                 coffee   
1     alternative energy           clean energy         climate change   
2                    NaN     alternative energy            electricity   
3           clean energy  community empowerment  rural electrification   
4             innovation     polution reduction           solar energy   

           Thematic Tag4 Thematic Tag5         Lat        Long        image_1  \
0                 olives           NaN  36.7607349   3.0261856            NaN   
1  community empowerment           NaN  -12.833226  13.6010742   file: img-10   
2     polution reduction           NaN  -8.5158356  16.0817871  file: img-121   
3         social justice       welfare  -12.776115   15.687688  file: img-230   
4                    NaN           NaN   -12.86536  13.8098145  file: img-305   

  image_2 image_3 image_

In [25]:
energy_data['Clean Cooking'].nunique(dropna=False)

2

In [26]:
energy_data[['Clean Cooking', 'Open Source', 'IP Protected', 'Can Train Others',
       'Prototype', 'Product', 'Market Ready',
       'Requires Advanced Order']].describe()

Clean Cooking  Open Source  IP Protected Can Train Others  Prototype  \
count            359          100            41              359         69   
unique             2            1             1                2          1   
top            False  Open Source  IP Protected            False  Prototype   
freq             277          100            41              232         69   

        Product Market Ready Requires Advanced Order  
count       124          359                     359  
unique        1            2                       2  
top     Product        False                   False  
freq        124          295                     298

In [27]:
energy_data.isnull().sum()

Date                         2
ID                           0
Energy Source                4
Clean Cooking                0
Project Title                0
Mapper                       0
Contributor                  0
Purpose of Solution         12
Link                       233
Open Source                259
IP Protected               318
Can Train Others             0
Unit Cost in USD           289
Prototype                  290
Product                    235
Market Ready                 0
Requires Advanced Order      0
Tech Readiness Level       233
User Feedback              256
Feedback Link              332
Efficiency Benchmark       301
Potential Bottlenecks      306
SDG Solution1               34
SDG Solution2              122
SDG Solution3              181
SDG Solution4              284
SDG Solution5              341
Thematic Tag1               39
Thematic Tag2               77
Thematic Tag3              146
Thematic Tag4              235
Thematic Tag5              314
Lat     

In [28]:
energy_data['Thematic Tag1'].nunique()

71

In [29]:
energy_data[energy_data['Mapper'] == 'AccLab Panama']

Date    ID          Energy Source  Clean Cooking  \
226 2022-05-17 21:45:44  4784  Household application           True   
227 2022-05-17 22:20:08  4785                  Solar          False   
228 2022-05-24 00:17:39  4818                Thermal          False   
229 2022-05-24 00:44:04  4819                Thermal          False   
230 2022-05-24 01:03:40  4820                Thermal          False   
231 2022-05-24 01:11:53  4821                Thermal          False   
232 2022-05-24 01:18:03  4822                Thermal          False   
233 2022-05-24 01:21:48  4823                Thermal          False   
234 2022-05-24 01:26:59  4824                Thermal          False   
235 2022-05-24 01:32:54  4825                Thermal          False   
236 2022-05-24 01:37:42  4826                  Solar          False   
237 2022-05-24 01:45:15  4827                Thermal          False   
238 2022-05-24 01:48:57  4828                Thermal          False   
239 2022-05-24 01:56:27  4829                Thermal          False   
240 2022-05-25 04:38:51  4835             Mechanical          False   
241 2022-05-25 06:14:04  4837  Household application           True   
242 2022-05-25 06:43:48  4838                 Solar           False   
243 2022-05-25 07:40:18  4839                  Solar          False   
244 2022-05-27 04:41:00  4899      Renewable general          False   
245 2022-05-27 04:59:50  4900                  Solar          False   
246 2022-05-27 05:07:42  4901                  Hydro          False   
247 2022-05-27 05:14:07  4902                  Solar          False   
248 2022-05-27 05:35:16  4904      Renewable general          False   
249 2022-05-27 05:52:14  4905                  Solar          False   
250 2022-05-27 06:14:50  4907      Renewable general          False   
251 2022-05-28 02:07:17  4922      Renewable general          False   
252 2022-05-28 02:16:47  4923      Renewable general          False   
253 2022-06-08 09:16:36  5048      Renewable general          False   

                                       Project Title         Mapper  \
226                   Biodigestors for clean cooking  AccLab Panama   
227                                        IluPanama  AccLab Panama   
228             Thermoelectric Plant - Isla Chepillo  AccLab Panama   
229             Thermoelectric plant - Otoque Island  AccLab Panama   
230           Thermoelectric Plant - Isla San Miguel  AccLab Panama   
231           Thermoelectric - Pedro Gonzalez Island  AccLab Panama   
232                   Thermoelectric - Taboga Island  AccLab Panama   
233                   Thermoelectric - Saboga Island  AccLab Panama   
234                         Thermoelectric - Nargana  AccLab Panama   
235                 Themoelectric Plant - Ro Az√∫car  AccLab Panama   
236                           Hybrid Plant - La Miel  AccLab Panama   
237             Thermoelectric Plant - Puerto Obalda  AccLab Panama   
238  Thermoelectric Plants - Eastern Panama & Darien  AccLab Panama   
239                Thermoelectric Plant - Isla Colon  AccLab Panama   
240                                       Maya Pedal  AccLab Panama   
241                                   Estufas Chispa  AccLab Panama   
242                               Amigos de la Aldea  AccLab Panama   
243                                           KINTEK  AccLab Panama   
244                                         Ecocomal  AccLab Panama   
245                                  Paneles Solares  AccLab Panama   
246                               Calentador de agua  AccLab Panama   
247          Paneles solares con gestion comunitaria  AccLab Panama   
248   Programa de Resiliencia Energetica Comunitaria  AccLab Panama   
249              FUJEM 2020 (Fundacion Mujeres 2020)  AccLab Panama   
250                           Energa Eolica e Hdrica  AccLab Panama   
251                             Iluminando tu Futuro  AccLab Panama   
252                                       

In [30]:
energy_data[energy_data['Lat'] == 'xxxx']

Date    ID Energy Source  Clean Cooking  \
200 2021-08-24 10:32:22  3458       Thermal          False   

                         Project Title        Mapper      Contributor  \
200  Alttakaddum ou charbon ecologique  AccLab Niger  Contributor 201   

                                   Purpose of Solution  \
200  Cette innovation consiste a transformer les ca...   

                                                  Link Open Source  \
200  https://m.facebook.com/story.php?story_fbid=10...         NaN   

    IP Protected  Can Train Others Unit Cost in USD Prototype Product  \
200          NaN             False              NaN       NaN     NaN   

     Market Ready  Requires Advanced Order Tech Readiness Level User Feedback  \
200         False                    False                  NaN           NaN   

    Feedback Link Efficiency Benchmark Potential Bottlenecks  SDG Solution1  \
200           NaN                  NaN                   NaN            1.0   

     SDG Solution2  SDG Solution3  SDG Solution4  SDG Solution5  \
200            3.0           12.0           13.0            NaN   

          Thematic Tag1 Thematic Tag2               Thematic Tag3  \
200  alternative energy  clean energy  protecting the environment   

           Thematic Tag4     Thematic Tag5   Lat   Long        image_1  \
200  recycle waste plant  waste management  xxxx  xxxxx  file: img-377   

    image_2 image_3 image_4 image_5 image_6 image_7 image_8 image_9  
200     NaN     NaN     NaN     NaN     NaN     NaN     NaN     NaN

In [31]:
# Replace xxxx and xxxxx with nan in both columns respectively
energy_data['Lat'] = energy_data['Lat'].replace('xxxx', np.nan)
energy_data['Long'] = energy_data['Long'].replace('xxxxx', np.nan)

In [32]:
energy_data[energy_data['Lat'] == 'xxxx']

Empty DataFrame
Columns: [Date, ID, Energy Source, Clean Cooking, Project Title, Mapper, Contributor, Purpose of Solution, Link, Open Source, IP Protected, Can Train Others, Unit Cost in USD, Prototype, Product, Market Ready, Requires Advanced Order, Tech Readiness Level, User Feedback, Feedback Link, Efficiency Benchmark, Potential Bottlenecks, SDG Solution1, SDG Solution2, SDG Solution3, SDG Solution4, SDG Solution5, Thematic Tag1, Thematic Tag2, Thematic Tag3, Thematic Tag4, Thematic Tag5, Lat, Long, image_1, image_2, image_3, image_4, image_5, image_6, image_7, image_8, image_9]
Index: []

In [33]:
energy_data[energy_data['Long'] == 'xxxxx']

Empty DataFrame
Columns: [Date, ID, Energy Source, Clean Cooking, Project Title, Mapper, Contributor, Purpose of Solution, Link, Open Source, IP Protected, Can Train Others, Unit Cost in USD, Prototype, Product, Market Ready, Requires Advanced Order, Tech Readiness Level, User Feedback, Feedback Link, Efficiency Benchmark, Potential Bottlenecks, SDG Solution1, SDG Solution2, SDG Solution3, SDG Solution4, SDG Solution5, Thematic Tag1, Thematic Tag2, Thematic Tag3, Thematic Tag4, Thematic Tag5, Lat, Long, image_1, image_2, image_3, image_4, image_5, image_6, image_7, image_8, image_9]
Index: []

In [34]:
energy_data[['Lat', 'Long']] = energy_data[['Lat', 'Long']].astype('float64')

In [35]:
energy_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 359 entries, 0 to 358
Data columns (total 43 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Date                     357 non-null    datetime64[ns]
 1   ID                       359 non-null    int64         
 2   Energy Source            355 non-null    object        
 3   Clean Cooking            359 non-null    bool          
 4   Project Title            359 non-null    object        
 5   Mapper                   359 non-null    object        
 6   Contributor              359 non-null    object        
 7   Purpose of Solution      347 non-null    object        
 8   Link                     126 non-null    object        
 9   Open Source              100 non-null    object        
 10  IP Protected             41 non-null     object        
 11  Can Train Others         359 non-null    bool          
 12  Unit Cost in USD         70 non-null

In [36]:
# Reshape columns 'Prototype' and 'Product' to long form
id_vars = energy_data.drop(columns = ['Prototype', 'Product']).columns

energy_data = energy_data.melt(id_vars = id_vars, value_name = 'Project Stage')
energy_data = energy_data.drop(columns=['variable'])
energy_data.sample(25)

Date    ID          Energy Source  Clean Cooking  \
255 2022-05-16 13:43:21  4744                  Solar          False   
306 2020-09-22 00:14:24  1163      Renewable general          False   
541 2020-10-30 12:24:34  2119                  Solar          False   
17  2022-05-03 15:56:24  4590                  Solar          False   
117 2022-03-28 15:16:28  4346                  Solar          False   
350 2022-06-09 10:04:42  5084  Household application           True   
509 2022-08-01 11:56:43  5375  Household application          False   
363 2021-07-21 10:12:39  3236  Household application           True   
427 2022-06-07 19:33:47  5042                 Solar           False   
469 2022-05-04 08:35:02  4593                  Solar          False   
132 2022-06-02 17:54:28  4972                    NaN          False   
431 2022-06-10 12:15:09  5142  Household application           True   
382 2022-05-16 12:54:21  4742  Household application           True   
305 2021-01-14 10:26:59  2690                Thermal          False   
498 2022-01-28 18:28:23  4017      Renewable general          False   
190 2020-10-28 13:56:14  2103                  Solar          False   
93  2022-05-23 16:46:43  4816  Household application          False   
451 2022-05-23 16:12:09  4811                  Hydro          False   
635 2020-10-11 17:35:05  1795  Household application          False   
82  2020-07-14 11:52:39    21                Thermal          False   
708 2022-06-09 10:03:11  5083  Household application           True   
417 2022-06-09 16:34:58  5113                 Solar           False   
189 2020-09-10 13:44:00   146                Thermal          False   
346 2022-06-09 09:54:29  5079  Household application           True   
109 2022-04-13 13:48:43  4511      Renewable general          False   

                                         Project Title  \
255                              JM JULIEN DUSHIMIMANA   
306                      Theocentric Business solution   
541                                 Solar Photovaltaic   
17   A device that injects solar energy into the co...   
117                                        Future pump   
350                              2 in 1 firewood stove   
509                             Homemade Giant Battery   
363                                       Solar Cooker   
427          CLEAN TECH SOLAR BOX BYGADGETS INDUSTRIES   
469                         Manual sun tracking for PV   
132                                     Estufas Chispa   
431                            TEGOMO GAZ INSTANTANÉE   
382  Ecological kitchens and ovens, solar heaters, ...   
305                            SMART ENERGY GENERATION   
498                 Eco, energy-saving ways of cooking   
190                           Solar-Powered Hydroponic   
93                                  Smart Energy Meter   
451                  Mini hydraulics to produce energy   
635                                  Choose your Light   
82                         Charcoal lamp with USB port   
708   Natural Induction Stove ( Modified Wooden Stove)   
417                                Chauffe-eau solaire   
189                        Plastic Pyrolysis into Fuel   
346                       Pressure Type Kerosene Stove   
109                   100% renewable energy university   

                        Mapper      Contributor  \
255              AccLab Rwanda  Contributor 256   
306              AccLab Zambia  Contributor 307   
541            AccLab Malaysia  Contributor 183   
17            AccLab Argentina   Contributor 18   
117            AccLab Ethiopia  Contributor 118   
350  Honey Bee Network (India)  Contributor 351   
509              AccLab Guinea  Contributor 151   
363              AccLab Angola    Contributor 5   
427            AccLab Cameroon   Contributor 69   
469               AccLab Egypt  Contributor 111   
132           AccLab Guatemala  Contributor 133   
431            AccLab Cameroon   Contributor 73   


In [37]:
energy_data[['ID', 'Project Stage']].sample(20)

ID Project Stage
416  5112       Product
131  4971           NaN
644  5202       Product
407  1560           NaN
464  4353           NaN
84     20           NaN
670  1097           NaN
101   474           NaN
202  4627           NaN
156   151           NaN
36   4564           NaN
304  1128           NaN
589  4820       Product
505  5154       Product
419  5156           NaN
154    76           NaN
82     21           NaN
488  1554           NaN
119  4356           NaN
490  4971       Product

In [38]:
# Reshape columns 'Open Source' and 'IP Protected' to long form
id_vars = energy_data.drop(columns = ['Open Source', 'IP Protected']).columns

energy_data = energy_data.melt(id_vars = id_vars, value_name = 'Intellectual Property')
energy_data = energy_data.drop(columns=['variable'])
energy_data.sample(25)

Date    ID          Energy Source  Clean Cooking  \
933  2022-05-11 22:10:02  4679                  Solar          False   
969  2022-05-28 02:07:17  4922      Renewable general          False   
736  2022-05-03 16:24:29  4591                 Solar           False   
45   2022-06-22 17:52:37  5305                Thermal          False   
1212 2022-06-02 18:14:14  4975                  Solar          False   
480  2022-03-30 15:40:12  4359  Household application           True   
1238 2020-09-17 19:36:09   795      Renewable general          False   
1287 2022-05-10 23:08:29  4669                  Solar          False   
1049 2022-06-09 07:27:44  5054                  Hydro          False   
1133 2022-06-09 15:21:37  5111                  Solar          False   
1209 2022-06-02 17:54:28  4972                    NaN          False   
206  2022-05-10 21:28:15  4665                  Solar          False   
456  2022-06-07 15:25:14  5026  Household application           True   
1078 2021-06-28 13:56:30  3177                   Wind          False   
236  2022-05-24 01:37:42  4826                  Solar          False   
488  2020-09-24 13:21:39  1554               Chemical          False   
937  2022-05-11 23:58:51  4683  Household application           True   
1277 2021-08-24 10:32:22  3458                Thermal          False   
572  2022-05-11 21:29:10  4677                  Solar          False   
829  2020-10-11 11:39:53  1780  Household application          False   
468  2022-04-13 13:48:43  4511      Renewable general          False   
727  2022-04-22 15:06:22  4546                  Solar          False   
910  2022-04-05 22:35:03  4374      Renewable general          False   
1338 2020-10-07 08:23:43  1707                 Solar           False   
672  2020-09-21 13:51:43  1139                  Solar          False   

                                          Project Title  \
933   Largest marina solar system in the world—ins...   
969                                Iluminando tu Futuro   
736   Prototype that "captures" sunlight and transmi...   
45    Fuel from sargassum, rum wastewater & sheep ma...   
1212                                 Amigos de la Aldea   
480                             Lump charcoal cookstove   
1238                                    Watering System   
1287                            Solar-powered long boat   
1049                     Turbinator (Turbine Generator)   
1133                             Foyer ameliore hybride   
1209                                     Estufas Chispa   
206                             Solar UV drier for kava   
456                                 Eco-efficient stove   
1078                               Wind power generator   
236                              Hybrid Plant - La Miel   
488                          Alchemy Alternative Energy   
937                      Locally designed peanut cooker   
1277                  Alttakaddum ou charbon ecologique   
572                    Solar energy for off-grid living   
829                                   Portable Lighting   
468                    100% renewable energy university   
727   Solar Electric Mobility Prototype: EcoAndina F...   
910       Electric car fully assembled by Mexican women   
1338   Solar and Windmill Hybrid Electrification System   
672   Deaftronics building solar powered batteries f...   

                         Mapper      Contributor  \
933              AccLab Pacific  Contributor 216   
969               AccLab Panama  Contributor 252   
736            AccLab Argentina   Contributor 19   
45              AccLab Barbados   Contributor 46   
1212           AccLab Guatemala  Contributor 136   
480             AccLab Ethiopia  Contributor 122   
1238                AccLab Iraq  Contributor 162   
1287             AccLab Pacific  Contributor 211   
1049  Honey Bee Network (India)  Contributor 332   
1133        AccLab Burkina Faso   Contributor 57   
1209           AccLab Guatemala  Contrib

In [40]:
energy_data.shape

(1436, 41)

In [61]:
to_translate = "COUVEUSE PETRO-SOLAIRE M'YAABA"
import timeit

%timeit ts.google(to_translate, to_language = 'en')

1.38 s ± 364 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [62]:
# Translate columns to English
energy_data['Project Title'] = energy_data['Project Title'].astype('str').apply(lambda x : ts.google(x, to_language='en', 
                                                                                                     if_ignore_limit_of_length=True, 
                                                                                                    timeout = 1.38))

ConnectTimeout: HTTPSConnectionPool(host='translate.google.com', port=443): Max retries exceeded with url: / (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x000002570FCDB520>, 'Connection to translate.google.com timed out. (connect timeout=1.38)'))

In [ ]:
energy_data['Project Title']

In [44]:
french = 'bonjour'